rust/src/client.rs

```diff
@@ -0,0 +1,842 @@
+                let mut buf = Vec::<c_char>::with_capacity(1024);
+                let status = settings_read_str(
+                    self.inner.ctx,
+                    cgroup.as_ptr(),
+                    cname.as_ptr(),
+                    buf.as_mut_ptr(),
+                    buf.capacity().try_into().unwrap(),
+                );
+                (
+                    SettingValue::String(
+                        CStr::from_ptr(buf.as_ptr()).to_string_lossy().into_owned(),
+                    ),
+                    status,
+                )
+            },
+        };
+
+        debug!("{} {} {:?} {}", setting.group, setting.name, res, status);
+
+        if status == 0 {
+            Some(Ok(res))
+        } else {
+            Some(Err(status.into()))
+        }
+    }
+
+    pub fn write_setting(
+        &self,
+        group: &str,
+        name: &str,
+        value: &str,
+    ) -> Result<(), Error<WriteSettingError>> {
+        let cgroup = CString::new(group)?;
+        let cname = CString::new(name)?;
+        let cvalue = CString::new(value)?;
+        let mut event = Event::new();
+
+        let result = unsafe {
+            settings_write_str(
+                self.inner.ctx,
+                &mut event as *mut Event as *mut _,
+                cgroup.as_ptr(),
+                cname.as_ptr(),
+                cvalue.as_ptr(),
+            )
+        };
+
+        #[allow(non_upper_case_globals)]
+        match result {
+            settings_write_res_e_SETTINGS_WR_OK => Ok(()),
+            code => Err(code.into()),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Error<E> {
+    NulError,
+    Err(E),
+}
+
+impl<E> From<ffi::NulError> for Error<E> {
+    fn from(_: ffi::NulError) -> Self {
+        Error::NulError
+    }
+}
+
+impl<E> From<u32> for Error<E>
+where
+    E: From<u32>,
+{
+    fn from(err: u32) -> Self {
+        Error::Err(err.into())
+    }
+}
+
+impl<E> From<i32> for Error<E>
+where
+    E: From<i32>,
+{
+    fn from(err: i32) -> Self {
+        Error::Err(err.into())
+    }
+}
+
+impl<E> std::fmt::Display for Error<E>
+where
+    E: std::error::Error,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Error::NulError => write!(f, "error converting to CString"),
+            Error::Err(err) => write!(f, "{}", err),
+        }
+    }
+}
+
+impl<E> std::error::Error for Error<E> where E: std::error::Error {}
+
+#[derive(Debug, Clone)]
+pub struct ReadByIdxResult {
+    pub group: String,
+    pub name: String,
+    pub value: String,
+    pub fmt_type: String,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum WriteSettingError {
+    ValueRejected,
+    SettingRejected,
+    ParseFailed,
+    ReadOnly,
+    ModifyDisabled,
+    ServiceFailed,
+    Timeout,
+    Unknown,
+}
+
+impl From<i32> for WriteSettingError {
+    fn from(n: i32) -> Self {
+        (n as u32).into()
```

Comment:
~Does this panic if n is too large of a value?~ Code review nit: the `impl From<u32> for WriteSettingError {` should go before this one in top to bottom reading order 😅